### PR TITLE
Refactoring attention.py part 1

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

`attention.py` has grown in size to 8500+ lines of code and hence the motivation to refactor the code. This PR is part 1 (of a 2 part PR effort) to refactor `attention.py` into sub-modules for ease of development and testing.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

Create a new module, `dot_product_attention` which contains sub modules `inference.py`, `rope.py` and `utils.py` and move the appropriate methods from `attention.py` into these 3 sub modules. The details of these methods are in a collapsed list below for those interested: 

<details>

<summary>List of Functions Moved/Added</summary>

- rope.py (~200 lines of code)
class RotaryPositionEmbedding(torch.nn.Module)
class FusedRoPEFunc(torch.autograd.Function):
def _rotate_half(x: torch.Tensor) -> torch.Tensor:
def apply_rotary_pos_emb(

- inference.py (estimated to be ~1000 when finished)
class InferenceParams: # pylint: disable=too-few-public-methods

- utils.py (~1500 lines of code)
class AttentionParams: Agree with a different class. Unsure of it being in utils.py
def get_attention_backend(
def get_full_mask(
def get_alibi(
def get_cu_seqlens(mask: torch.Tensor) -> torch.Tensor:
def get_cu_seqlens_and_indices(mask: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
def get_indices(max_seqlen: int, cu_seqlens: torch.Tensor) -> torch.Tensor:
def _get_full_cu_seqlens(
def pack_tensor(
def pack_2_tensors(
def pack_3_tensors(
def unpack_tensor(
def unpack_2_tensors(
def unpack_3_tensors(
class PackTensors(torch.autograd.Function):
class UnpackTensor(torch.autograd.Function):
def get_qkv_layout(
def check_set_window_size(
def get_attention_quantizers - Added new clas
Create AttentionLogging class - Added new class
Create FlashAttentioUtils class - Added  new class
</details>

Notable function signature changes(comments have been added for these in the respective functions) :
1. `get_attention_backend()` - This will not populate the global `_attention_backends` cache now but rather rest the responsibility of populating the global `_attention_backends` on the caller of this function
2. `get_alibi()` - This will now accept the global `_alibi_cache` as a function parameter from the caller and read/write to the `_alibi_cache` cache from within the function

**TODO:** Refactoring part 2 PR will create new modules/submodules for `MultiHeadAttention.py`, `context_parallelism` while moving some more generic pytorch utils functions to `pytorch/utils.py` and also moving whatever is leftover of `attention.py` into `dot_product_attention` . 
Additionally, the part 2 PR will also address any larger changes suggested as part of this part 1 PR's review process.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by cyanguwa** [comment]
I wonder if it makes sense to leave those flash-attn version checks/imports as is, for now. We'd have to think of a cleaner way to do it, but I feel `FlashAttentionUtils` might not cut it.

Also, could you have a look at the usage of `InferenceParams` and RoPE, see if it makes more sense to keep them in `dot_product_attention/` or outside that directory? Thanks!

**Comment by yaox12** [comment]
> I wonder if it makes sense to leave those flash-attn version checks/imports as is, for now. We'd have to think of a cleaner way to do it, but I feel `FlashAttentionUtils` might not cut it.
> 
> Also, could you have a look at the usage of `inference.py` and `rope.py`, see if it makes more sense to keep them in `dot_product_attention/` or outside that directory? Thanks!

I'm thinking if we should create a folder naming like `transformer_engine.pytorch.functional` (similar to `torch.nn.functional`) for ops but not layers. And move RoPE, [softmax](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/softmax.py), [permutation](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/permutation.py), [cross_entropy](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/cross_entropy.py) to it. How do you like it? @cyanguwa

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -87,71 +74,45 @@
     restore_from_saved,
 )
 
+# Import attention utils
+import transformer_engine.pytorch.dot_product_attention.utils as dpa_utils
+import transformer_engine.pytorch.dot_product_attention.inference as dpa_infer
+from transformer_engine.pytorch.dot_product_attention.utils import FlashAttentionUtils as fa_utils
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:80]: `FlashAttentionUtils` is probably a better name than `FAUtils`. Thanks.

File: transformer_engine/pytorch/attention.py
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:1]: I think this probably is a good structure for our finished refactoring (maybe not now):
```
te/pytorch:
  - transformer.py: TELayer (~800 LOC)
  - multihead_attention.py: MHA (~800 LOC)
  - attention/
     - attention.py: DPA (~1000 LOC)
     - backends.py: FusedAttention, FlashAttention, UnfusedDPA (~1200 LOC)
     - context_parallel.py: P2P, A2A, AllGather, attn_with_cp (~3300 LOC)
     - softmax.py: only used in attention.py IIUC
     - utils.py: ~1600 LOC, but can move _SplitAlongDim, _combine_tesnsors to te/pytorch/utils.py (they can probably be merged with noop_cat, need investigation), and can condense PackTensors/UnpackTensors related funcs (one pack_tensors func instead of 3)
  - rope.py: ~200 LOC, but might get longer (a couple of PRs in the pipeline)
  - inference.py: ~50 LOC, but will get longer (PR 1355, 800 LOC)
```

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -87,71 +74,45 @@
...
-_flash_attn_2_1_plus = False
-_flash_attn_2_3_plus = False
-_flash_attn_2_4_plus = False
-_flash_attn_2_4_1_plus = False
-_flash_attn_2_5_7_plus = False
-_flash_attn_2_6_0_plus = False
-_flash_attn_2_7_0_plus = False
 
+# Setup Attention Logging
+att_log.setup_logging()
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:86]: Nit: I think people usually abbreviate attention as attn.

**Review by cyanguwa — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/attention.py
Comment (KshitijLakhani) [reply transformer_engine/pytorch/attention.py:1]: Sounds good. 
Will address it in the part 2 PR. Added to the TODO list as well

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py:1]: I feel that if we're keeping `multihead_attention.py` outside, we should rename the dir structure as follows:
```
te/pytorch:
  - transformer.py: TELayer (~800 LOC)
  - multihead_attention.py: MHA (~800 LOC)
  - dot_product_attention/
     - dot_product_attention.py: DPA (~1000 LOC)
     - backends.py: FusedAttention, FlashAttention, UnfusedDPA (~1200 LOC)
     - context_parallel.py: P2P, A2A, AllGather, attn_with_cp (~3300 LOC)
     - softmax.py: only used in attention.py IIUC
     - utils.py: ~1600 LOC, but can move _SplitAlongDim, _combine_tesnsors to te/pytorch/utils.py (they can probably be merged with noop_cat, need investigation), and can condense PackTensors/UnpackTensors related funcs (one pack_tensors func instead of 3)
  - rope.py: ~200 LOC, but might get longer (a couple of PRs in the pipeline)
  - inference.py: ~50 LOC, but will get longer (PR 1355, 800 LOC)
```
Or keep `multihead_attention.py` inside attention as follows. Keeping `m_h_a.py` as a file top level in addition to attention as directory with attention/attention.py as a file could be misleading.
```
te/pytorch:
  - transformer.py: TELayer (~800 LOC)
  - attention/
     - multihead_attention.py: MHA/GQA/MLA
     - dot_product_attention.py: DPA (~1000 LOC)
     - backends.py: FusedAttention, FlashAttention, UnfusedDPA (~1200 LOC)
     - context_parallel.py: P2P, A2A, AllGather, attn_with_cp (~3300 LOC)
     - softmax.py: only used in attention.py IIUC
     - utils.py: ~1600 LOC, but can move _SplitAlongDim, _combine_tesnsors to te/pytorch/utils.py (they can probably be merged with noop_cat, need investigation), and can condense PackTensors/UnpackTensors related funcs (one pack_tensors func instead of 3)
  - rope.py: ~200 LOC, but might get longer (a couple of PRs in the pipeline)
  - inference.py: ~50 LOC, but will get longer (PR 1355, 800 LOC)
```

The latter looks cleaner to me, if doesn't break more stuff than the former

**Review by sudhakarsingh27 — COMMENTED** [review]
Overall lgtm since this is just setting up for Part 2

File: transformer_engine/pytorch/dot_product_attention/rope.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
+# Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+"""
+Rotary Position Embedding implementation of different types along with hlper functions
```
Comment (cyanguwa) [inline transformer_engine/pytorch/dot_product_attention/rope.py:6]: Nit: "helper"

File: transformer_engine/pytorch/dot_product_attention/utils.py [hunk 4]
```diff
@@ -0,0 +1,1641 @@
...
+
+# ----Global constants----
+# NVTE_DEBUG = 0/1 # disables/enables debug mode, default = 0
+_NVTE_DEBUG = int(os.getenv("NVTE_DEBUG", "0"))
+# NVTE_DEBUG_LEVEL = 0/1/2 # enables more and more verbose debug mode, default = 0
+_NVTE_DEBUG_LEVEL = int(os.getenv("NVTE_DEBUG_LEVEL", "0"))
+_NVTE_FLASH_ATTN = int(os.getenv("NVTE_FLASH_ATTN", "1"))
+
+
+# ----Helper/Util classes and methods-----
```
Comment (cyanguwa) [inline transformer_engine/pytorch/dot_product_attention/utils.py:57]: Nit: maybe remove these comments? L57 and L49? It doesn't look like we make these comments elsewhere in utils.py.

File: transformer_engine/pytorch/attention.py [hunk 5]
```diff
@@ -87,71 +74,45 @@
     restore_from_saved,
 )
 
+# Import attention utils
+import transformer_engine.pytorch.dot_product_attention.utils as dpa_utils
+import transformer_engine.pytorch.dot_product_attention.inference as dpa_infer
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:79]: Nit: could probably do "from transformer_engine.pytorch.dot_product_attention.inference import InferenceParams"?

**Review by cyanguwa — APPROVED** [review]
LGTM. After all CI passes, you can merge.

**Comment by cyanguwa** [comment]
> > Also, could you have a look at the usage of `inference.py` and `rope.py`, see if it makes more sense to keep them in `dot_product_attention/` or outside that directory? Thanks!
> 
> I'm thinking if we should create a folder naming like `transformer_engine.pytorch.functional` (similar to `torch.nn.functional`) for operators (here I mean what we usually use as autograd functions, not through layers or modules). And move RoPE, [softmax](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/softmax.py), [permutation](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/permutation.py), [cross_entropy](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/cross_entropy.py) to it. How do you like it? @cyanguwa

Let's have a think about this in Refactoring Part 2. @KshitijLakhani
